src/compas/robots/base_artist/_artist.py

```diff
@@ -0,0 +1,309 @@
+    of the CAD environment.
+
+    There are two methods that implementers of this base class should provide, one
+    is concerned with the actual creation of geometry in the native format of the
+    CAD environment (:meth:`draw_geometry`) and the other is one to apply a transformation
+    to geometry (:meth:`transform`).
+
+    Attributes
+    ----------
+    model : :class:`compas.robots.RobotModel`
+        Instance of a robot model.
+    """
+
+    def __init__(self, model):
+        super(BaseRobotModelArtist, self).__init__()
+        self.model = model
+        self.create()
+        self.scale_factor = 1.
+        self.attached_tool_model = None
+
+    def attach_tool_model(self, tool_model):
+        """Attach a tool to the robot artist.
+
+        Parameters
+        ----------
+        tool_model : :class:`compas.robots.ToolModel`
+            The tool that should be attached to the robot's flange.
+        """
+        self.attached_tool_model = tool_model
+        self.create(tool_model.root, 'attached_tool')
+
+        if not tool_model.link_name:
+            link = self.model.get_end_effector_link()
+            tool_model.link_name = link.name
+        else:
+            link = self.model.get_link_by_name(tool_model.link_name)
+
+        # originally, there was only this
+        ee_frame = link.parent_joint.origin.copy()
+        initial_transformation = Transformation.from_frame_to_frame(Frame.worldXY(), ee_frame)
+        # but if you attach a tool after applying some update to the robot, then it appears
+        # disconnected, being displayed at the original location of the end effector frame.
+        # I feel it shouldn't matter the order of calling attach_tool(_model) and update
+        # so I suggest the alternative, but maybe it isn't safe.
+
+        if link.collision:
+            relative_transformation = link.collision[0].current_transformation
+        else:
+            relative_transformation = link.visual[0].current_transformation
+
+        transformation = relative_transformation.concatenated(initial_transformation)
+
+        self.update_tool(transformation=transformation)
+
+        tool_model.parent_joint_name = link.parent_joint.name
+
+    def detach_tool_model(self):
+        """Detach the tool.
+        """
+        self.attached_tool_model = None
+
+    def create(self, link=None, name=None):
+        """Recursive function that triggers the drawing of the robot model's geometry.
+
+        This method delegates the geometry drawing to the :meth:`draw_geometry`
+        method. It transforms the geometry based on the saved initial
+        transformation from the robot model.
+
+        Parameters
+        ----------
+        link : :class:`compas.robots.Link`, optional
+            Link instance to create. Defaults to the robot model's root.
+        name : :obj:`str`, optional
+            Subdomain identifier. # !!! there's got to be a better explanation.
```

Comment:
Suggestions for a better docstring?

Reply:
is `subdomain` referring to the fact that the name has to be unique within a certain context? if that is the case i would say something like "Human-readable identifier for the link. Note that the identifier is unique within ..."